homeassistant/components/homeconnect/api.py

```diff
@@ -0,0 +1,389 @@
+        except (HomeConnectError, ValueError):
+            _LOGGER.debug("Unable to fetch appliance status. Probably offline.")
+        try:
+            self.appliance.get_settings()
+        except (HomeConnectError, ValueError):
+            _LOGGER.debug("Unable to fetch settings. Probably offline.")
+        try:
+            program_active = self.appliance.get_programs_active()
+        except (HomeConnectError, ValueError):
+            _LOGGER.debug("Unable to fetch active programs. Probably offline.")
+            program_active = None
+        if program_active and "key" in program_active:
+            self.appliance.status["BSH.Common.Root.ActiveProgram"] = {
+                "value": program_active["key"]
+            }
+        self.appliance.listen_events(callback=self.event_callback)
+        self.entities = []
+
+    def event_callback(self, appliance):
+        """Handle event."""
+        _LOGGER.debug("Update triggered on %s", appliance.name)
+        _LOGGER.debug(self.entities)
+        _LOGGER.debug(self.appliance.status)
+        for entity in self.entities:
+            entity.async_entity_update()
+
+
+class HomeConnectEntity(Entity):
+    """Generic Home Connect entity (base class)."""
+
+    def __init__(self, device, name):
+        """Initialize the entity."""
+        self.device = device
+        self._name = name
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the node (used for Entity_ID)."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return the unique id base on the id returned by Home Connect and the entity name."""
+        return f"{self.device.appliance.haId}-{self.name}"
+
+    @property
+    def device_info(self):
+        """Return info about the device."""
+        return {
+            "identifiers": {(DOMAIN, self.device.appliance.haId)},
+            "name": self.device.appliance.name,
+            "manufacturer": self.device.appliance.brand,
+            "model": self.device.appliance.vib,
+        }
+
+    @callback
+    def async_entity_update(self):
+        """Update the entity."""
+        _LOGGER.debug("Entity update triggered on %s", self)
+        self.async_schedule_update_ha_state(True)
+
+
+class DeviceWithPrograms(HomeConnectDevice):
+    """Device with programs."""
+
+    _programs = []
+
+    def get_programs_available(self):
+        """Get the available programs."""
+        return self._programs
+
+    def get_program_switches(self):
+        """Get a dictionary with info about program switches."""
+        programs = self.get_programs_available()
+        return [{"device": self, "program_name": p["name"]} for p in programs]
+
+    def get_program_sensors(self):
+        """Get a dictionary with info about program sensors."""
+        sensors = {
+            "Remaining Program Time": "s",
```

Comment:
We should not have relative times in the state machine. Instead we should use the `timestamp` [device class](https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes) and make the state a UTC timestamp.

Reply:
This allows the user to have a countdown in their UI.

BTW, how many programs can be active at the same time? Just 1 right? 

Wouldn't it make more sense to have a start time, stop time, progress % and current program sensor ?

Reply:
Or am I misinterpreting something and this is just 1.

Reply:
Right, there is just one remaining time sensor, one progress sensor, and one elapsed time sensor per device/appliance. It's called "sensors" only because there are those 3. I should probably expand the doc string to make it clearer. Basically this maps to what the API gives us.